File: Plugins/Orimath.Basics.View/ViewModels/LineViewModel.cs
﻿using System;
using Orimath.Controls;
using Orimath.Core;
using Orimath.Plugins;

namespace Orimath.Basics.View.ViewModels
{
    public class LineViewModel : NotifyPropertyChanged, IDisplayTargetViewModel
    {
        private readonly ILayerModel? _layer;

        public LineViewModel(ILayerModel? layer, LineSegment line, IViewPointConverter pointConverter)
        {
            _layer = layer;
            Source = line;
            (X1, Y1) = pointConverter.ModelToView(line.Point1);
            (X2, Y2) = pointConverter.ModelToView(line.Point2);
        }

        public LineSegment Source { get; }
        public double X1 { get; }
        public double Y1 { get; }
        public double X2 { get; }
        public double Y2 { get; }
        public double XFactor => Source.Line.XFactor;
        public double YFactor => NearlyEquatable.Negate(Source.Line.YFactor);
        public double Intercept => Source.Line.Intercept;
        public double Slope => NearlyEquatable.UnaryPlus(XFactor / YFactor);
        public double Angle => NearlyEquatable.UnaryPlus(Math.Atan2(XFactor, YFactor) / Math.PI * 180.0) % 180.0;
        public double Length => Source.Length;

        public (ILayerModel, DisplayTarget) GetTarget()
        {
            if (_layer is null) throw new InvalidOperationException();
            return (_layer, DisplayTarget.NewLine(Source));
        }
        public override string ToString() => $"{Source.Line}\r\n傾き:{Slope:0.#####} 角度:{Angle:0.#####}°";
    }
}

File: Plugins/Orimath.Basics.View/ViewModels/LineViewModel.fs
namespace Orimath.Basics.View.ViewModels
open System
open Orimath.Core
open Orimath.Plugins
open Orimath.Controls
open Orimath.Basics.View
open Orimath.Core.NearlyEquatable
open Orimath.Combination

type LineViewModel(crease: Crease, pointConverter: IViewPointConverter, ?layer: ILayerModel) =
    inherit NotifyPropertyChanged()
    let (ScreenPoint(x1, y1)) = pointConverter.ModelToView(crease.Point1)
    let (ScreenPoint(x2, y2)) = pointConverter.ModelToView(crease.Point2)
    member _.Source = crease
    member _.X1 = x1
    member _.Y1 = y1
    member _.X2 = x2
    member _.Y2 = y2
    member _.XFactor = crease.Line.XFactor
    member _.YFactor = !-crease.Line.YFactor
    member _.Intercept = crease.Line.Intercept
    member _.Slope = !+(crease.Line.XFactor / crease.Line.YFactor)
    member _.Angle = !+(atan2 crease.Line.XFactor crease.Line.YFactor / Math.PI * 180.0) % 180.0
    member _.Length = crease.Length
    member _.Color =
        match crease.Type with
        | CreaseType.Draft -> InstructionColor.LightGray
        | _ -> InstructionColor.Black
   
    override _.ToString() = crease.ToString()

    interface IDisplayTargetViewModel with
        member _.GetTarget() =
            match layer with
            | Some(l) -> l, DisplayTarget.Crease(crease)
            | None -> invalidOp "Layer is not set"
